### ISSUE TITLE
file upload: change to copyfile

### ISSUE THREAD

**Issue by alpapan** [galaxyproject/galaxy#4527]
Hello

I successfully integrated owncloud with galaxy but I had to make a small change in the upload.py script.

(caveat I'm a newbie to post 2010 galaxy):

My data was failing because I was getting a permission denied for every file being uploaded via the Owncloud (as an FTP option). Shouldn't happen really as read permissions were fine. 

Turns out it was galaxy wanting to implement a chmod command on the original file (which is not owned by galaxy). And it really shouldn't do that...

The fix was as per [this stack overflow](https://stackoverflow.com/questions/11835833/why-would-shutil-copy-raise-a-permission-exception-when-cp-doesnt): i.e. use shutil.copyfile instead of shutil.copy.

cheers
alexie

**Comment by mvdbeek** [comment]
That sounds like a reasonable change to me, do you want to open a PR for this or should we handle this ?

**Comment by alpapan** [comment]
As I don't know which copy/copyfile to change without breaking galaxy, I will have to leave to you.

thanks
a

**Comment by mvdbeek** [comment]
I'm not sure I understand. So you didn't fix it ?
I tried this (with `ftp_upload_purge = False`), and I get the following traceback
```
Traceback (most recent call last):
  File "/Users/mvandenb/src/galaxy/tools/data_source/upload.py", line 428, in 
    __main__()
  File "/Users/mvandenb/src/galaxy/tools/data_source/upload.py", line 416, in __main__
    add_file(dataset, registry, json_file, output_path)
  File "/Users/mvandenb/src/galaxy/tools/data_source/upload.py", line 295, in add_file
    line_count, converted_path = sniff.convert_newlines(dataset.path, in_place=in_place, tmp_dir=tmpdir, tmp_prefix=tmp_prefix)
  File "/Users/mvandenb/src/galaxy/lib/galaxy/datatypes/sniff.py", line 136, in convert_newlines
    shutil.move(temp_name, fname)
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/shutil.py", line 302, in move
    copy2(src, real_dst)
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/shutil.py", line 130, in copy2
    copyfile(src, dst)
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/shutil.py", line 83, in copyfile
    with open(dst, 'wb') as fdst:
IOError: [Errno 13] Permission denied: u'/Users/mvandenb/src/galaxy/database/ftp/m.vandenbeek@gmail.com/test/test.txt'
```
Which can't be fixed by what you have suggested. I suppose the newline conversion should not happen in-place when `ftp_upload_purge` is set to false.

**Comment by alpapan** [comment]
Thanks. 
In your case it seems you had different user permissions than I (i had group write  for a group www and galaxy belong to so it proceeded fine).

So yes it seems that #4539 is the root of the problem so my suggestion may be a moot point.

However it seems to me that .copy should be used only when you are the owner of a file as that's required for permission changes (chmod). copyfile can be used when you are not (it doesn't do a chmod).

**Comment by alpapan** [comment]
Sorry, just switched to the DEV branch so i could test this. Yes it works perfect indeed and resolved by #4539 
thanks!
